olmo/eval/downstream.py

```diff
@@ -9,11 +9,62 @@
 from sklearn.metrics import f1_score
 from torchmetrics import Metric
 
+from olmo.torch_util import barrier, get_fs_local_rank
+from olmo.util import load_dataset_from_disk, save_dataset_to_disk
+
 from ..tokenizer import Tokenizer
 
 log = logging.getLogger(__name__)
 
 
+def load_dataset(path, name, split, datasets_cache_dir: Optional[str] = None):
+    # Try get dataset from disk.
+    if datasets_cache_dir is not None:
+        try:
+            return load_dataset_from_disk(path, name, split, datasets_cache_dir)
+        except FileNotFoundError:
+            log.info(
+                "Path %s name %s split %s not present in local dir %s, loading from online",
+                path,
+                name,
+                split,
+                datasets_cache_dir,
+            )
+        # Barrier here to stop the case where FS rank 0 saves the dataset to disk before some non-zero
+        # ranks try getting the dataset from disk. This would cause those non-zero ranks to bypass
+        # the next barrier and cause rank 0 to be stuck at that barrier.
+        barrier()
```

Comment:
This seems a little sketch, but I guess it's fine as long as all ranks either have or do not have the dataset on disk, which is probably a reasonable assumption?

Reply:
[9f2d9d2](https://github.com/allenai/OLMo/pull/623/commits/9f2d9d25aa61f749f545fbd1a058f70227838f8c)
I've changed the logic to be simpler and more robust against this case. Now the idea is that FS local rank 0 does its thing first (load from disk or online and cache), then every other rank follows afterwards. This is less optimized but simpler and safer imo.

Reply:
I like that much better ✅